### PR TITLE
Zoom Setup Wizard

### PR DESCRIPTION
**Pull request by mickmister**
#### Summary

This PR adds a setup wizard for the Zoom plugin.  An e2e test is written to test a basic walkthrough of the setup.

Server running the wizard https://zoom-seup-wizard.test.mattermost.cloud

#### Ticket Link

Fixes https://mattermost.atlassian.net/browse/MM-40936

#### Known todos

- [x] Telemetry needs to be added
- [x] Make it so webhook secret isn't exposed in post content. Should use interactive dialog to expose the information instead
- [ ] Check length of client id and client secret
- [ ] If the user runs the command in a channel other than the Zoom bot DM channel, it should respond with an ephemeral message pointing to the DM channel.

- [x] We should say to ignore "Verification Code" on the webhook event page
- [x] Change "Install" step to "Activation" in the finish step
- [x] Instead of saying "pick these scopes", say "add these scopes to the scopes already selected", because zoom now preselects scopes based on your webhook events

#### Known issues

- [x] Images are not showing fully until you click them to show in a modal. Maybe we need to use slack attachment images

### CODE REVIEW

File: server/wizard/steps/step_zoom_self_hosted.go
Comment (mickmister) [inline server/wizard/steps/step_zoom_self_hosted.go:18]: Need a better way of wording this, it's a bit clunky at the moment. The message to get across is:

> Yes or no, do you have a private zoom instance?

File: server/wizard/steps/step_webhook_events.go
Comment (mickmister) [inline server/wizard/steps/step_webhook_events.go:21]: We can use these in future features, and I'm wondering if we should ask admins to update the events later when we support a feature for them, or if we should have them preemptively set them to make it an easier transition when we support the event types.

File: server/wizard/steps/step_goto_zoom_marketplace.go
Comment (mickmister) [inline server/wizard/steps/step_goto_zoom_marketplace.go:23]: Not good wording. Probably should get rid of the title

File: server/wizard/steps/step_goto_zoom_marketplace.go
Comment (mickmister) [inline server/wizard/steps/step_goto_zoom_marketplace.go:17]: @aaronrothschild Any ideas on how I can get a self-hosted zoom server to test this out on?

Comment (larkox) [reply server/wizard/steps/step_goto_zoom_marketplace.go:17]: If I remember correctly, it is not exactly self hosted, but custom url. Like `companyName.zoom.us`. I think almost any company account is able to do this. I think (if you have the creds) you could use the mattermost one. If not, maybe aaron has anything else you can use.

File: server/wizard/steps/step_webhook_configuration.go
Comment (larkox) [inline server/wizard/steps/step_webhook_configuration.go:21]: Very dangerous TODO :sweat_smile: (it may slip into the final version)
Do you mind investigating what are the implications of marking/not marking this field and share with @aaronrothschild so he can make a decision on wording?

File: server/wizard/steps/step_webhook_configuration.go
Comment (larkox) [inline server/wizard/steps/step_webhook_configuration.go:20]: Same comment about inline code blocks for links.

File: server/wizard/steps/step_webhook_events.go
Comment (larkox) [reply server/wizard/steps/step_webhook_events.go:21]: I consider that we should ask only for what we use.
I understand the feeling, but I think just asking for what is needed will force us to consider the update path in future releases, which is a good thing.

File: server/wizard/steps/step_zoom_self_hosted.go
Comment (larkox) [reply server/wizard/steps/step_zoom_self_hosted.go:18]: Are we sure Zoom offer exactly that? I would check Zoom's own wording on this.

File: server/wizard/steps/step_greeting.go
Comment (larkox) [inline server/wizard/steps/step_greeting.go:23]: Just for my knowledge, what does it mean returning 0 or 999?

File: server/wizard/steps/step_create_zoom_app.go
Comment (larkox) [inline server/wizard/steps/step_create_zoom_app.go:15]: Shouldn't we have a step for Account level app and another step for User Level app?

File: tests-e2e/cypress/integration/zoom_setup_wizard.js
Comment (larkox) [inline tests-e2e/cypress/integration/zoom_setup_wizard.js:71]: You almost use no comments to explain what is going on, but in this particular comment, you are not using the "standard" that is stated at the header of the file.

File: tests-e2e/cypress/integration/zoom_setup_wizard.js
Comment (larkox) [inline tests-e2e/cypress/integration/zoom_setup_wizard.js:42]: Just for my knowledge: How do we know the test team exist and the zoom plugin is installed?

File: tests-e2e/cypress/integration/zoom_setup_wizard.js
Comment (larkox) [inline tests-e2e/cypress/integration/zoom_setup_wizard.js:112]: Corner case that we may or may not want to handle:
```
obj1 = {key1: 'something', key2: undefined}
obj2 = {key1: 'something', key3: 'something else'}
objectsEqual(obj1,obj2) // equals true
```

**Review by larkox — CHANGES_REQUESTED** [review]
Great work @mickmister ! A few changes here and there, but looks good in general.

**Comment by hanzei** [comment]
cc @dipak-demansol

**Comment by DHaussermann** [comment]
@mickmister I have down a round pf testing on one of the branches for Zoom and have some feedback. I also noticed it seems you have no screenshots yet some a captured appropriate ones while doing the test round. Let's follow-up to see if need them.

A. Can there be feedback when the users says "Not now" such as shoeing the setup command.
B. Is it suitable to ask about a pass a long to a Zoom admin here the way GitHub does it? 
C. If this question is Yes or No, can we clarify what type of Zoom is No? I can see admins not knowing the answer without explaining the common cloud based Zoom here 
image
D. Various missing images. See Dylan if you need screenshots.
E. Should we explicitly tell the admin they're making a user managed app?
F. I see values when I open the modal for the 1st time.

File: tests-e2e/cypress/integration/zoom_setup_wizard.js
Comment (mickmister) [reply tests-e2e/cypress/integration/zoom_setup_wizard.js:42]: Whoops I added the team creation to the GitHub plugin wizard e2e test but not the Zoom wizard https://github.com/mattermost/mattermost-plugin-github/blob/c22ad7188c35beddca3d3766e96e118b34bbeead/tests-e2e/cypress/integration/setup_wizard_spec.js#L39

@larkox The zoom plugin will be installed by CI, similar to how we're doing with the Apps plugin CI

File: server/wizard/steps/step_webhook_events.go
Comment (mickmister) [reply server/wizard/steps/step_webhook_events.go:21]: [We have discussed](https://community-daily.mattermost.com/core/pl/51mu7o7k9tnnjpbhwag9oaw48o) how we can tell admins of new features in plugin releases, though I don't think we've landed on a proper solution to implement. Maybe we can revive that discussion. @aaronrothschild What do you think?

I'm fine with removing these as it is incorrect to have them at the moment.

File: server/wizard/steps/step_greeting.go
Comment (mickmister) [reply server/wizard/steps/step_greeting.go:23]: @larkox @hanzei I was mainly following the conventions used in https://github.com/mattermost/mattermost-plugin-github/pull/527. I'm not entirely sure what each value means

File: server/wizard/steps/step_create_zoom_app.go
Comment (mickmister) [reply server/wizard/steps/step_create_zoom_app.go:15]: @larkox This effort is focused on getting set up quickly, so we are going with one method to keep the setup process simple. Linking to the documentation at the beginning and end of the flow will allow the admin to discover other options for set up.

**Comment by mickmister** [comment]
@DHaussermann For the password autocomplete issue you pointed out, can you please try out this mattermost-webapp branch to see if it is fixed there? `interactive-dialog-password-autocomplete-off`

**Comment by catalintomai** [comment]
@mickmister - do you need to integrate latest flow lib. updates: https://github.com/mattermost/mattermost-plugin-github/pull/533 ?

**Review by levb — APPROVED** [review]
the code LGTM, lint needs to get fixed but seems inconsequential. Leaving copy/testing to others.

File: server/http.go
Comment (hanzei) [inline server/http.go:76]: Nit pick: It idiomatic to write
```suggestion
	if state == "" {
```

Comment (hanzei) [inline server/http.go:76]: Checking if the state is empty might catch unwanted cases. Should the OAuth state be stored as part of the last wizard step?

File: server/telemetry.go
Comment (hanzei) [inline server/telemetry.go:52]:
```suggestion
func (p *Plugin) sendDailyTelemetry() {
	config := p.getConfiguration()
```

File: server/wizard/steps/step_create_zoom_app.go
Comment (hanzei) [inline server/wizard/steps/step_create_zoom_app.go:34]: Does it make sense to move the continue button into it's own function?

File: server/wizard/steps/step_greeting.go
Comment (hanzei) [inline server/wizard/steps/step_greeting.go:29]: Separate ticket?

File: server/wizard/steps/step_greeting.go
Comment (hanzei) [inline server/wizard/steps/step_greeting.go:12]: Should the introduction text be similar to the other plugin as first introduce the steps that will be taken?
Screenshot from 2022-02-25 12-45-15

File: server/wizard/steps/step_create_zoom_app.go
Comment (hanzei) [inline server/wizard/steps/step_create_zoom_app.go:29]: The other wizard have a `Cancel step` button. Should Zoom have the same for unification?

File: server/wizard/steps/step_webhook_configuration.go
Comment (hanzei) [inline server/wizard/steps/step_webhook_configuration.go:35]: This way the webhook secret will be stored in the post for every visible. In the JIRA wizard, an dialog is used to show confidential information like this.

File: server/wizard/steps/step_webhook_events.go
Comment (hanzei) [inline server/wizard/steps/step_webhook_events.go:18]: Can we verify that the webhook was successfully created by e.g. checking if an event was delivered?

File: server/wizard/wizard.go
Comment (hanzei) [inline server/wizard/wizard.go:77]: Separate ticket?

**Review by hanzei — CHANGES_REQUESTED** [review]
Great work Michael, especially on the E2E tests.

File: server/wizard/steps/step_zoom_self_hosted.go
Comment (mickmister) [reply server/wizard/steps/step_zoom_self_hosted.go]: @DHaussermann What is your usual workflow for testing this setting?

File: server/config/config.go
Comment (mickmister) [reply server/config/config.go:42]: @hanzei I'm not sure. IIRC you had concerns of lower/upper case changing when the admin is editing the config? What was your solution to this?

File: server/http.go
Comment (mickmister) [reply server/http.go:57]: I suppose not

File: server/http.go
Comment (mickmister) [reply server/http.go:76]: @hanzei This is a sort of fix for the case where the user clicks on a connect link without doing one of the following:

- Run `/zoom connect`
- Run `/zoom start`
- Click the channel header button to start a meeting

When one of these events occurs, the KV store is primed with info about if the user has the intention of starting a meeting after connection is complete, including what channel to send the "connected" ephemeral message to after connection is complete.

This block makes it so the user can just click the link without this priming. Then we default to the bot DM channel as the "follow up" channel. Basically, if you were to do this flow with current master, connecting wouldn't work due to lack of priming:

- `/zoom connect`
- Click connect link
- Do connection
- `/zoom disconnect`
- Click already existing connect link

So this new block makes it so you can connect with just the connect link.

File: server/wizard/steps/step_webhook_events.go
Comment (mickmister) [reply server/wizard/steps/step_webhook_events.go:18]: @hanzei I don't think zoom supports this. Do we want to include this as more of a manual process?

File: server/config/config.go
Comment (hanzei) [reply server/config/config.go:42]: My proposal would be unify and use lower cases everywhere. That means update the manifest.

File: server/wizard/steps/step_webhook_events.go
Comment (hanzei) [reply server/wizard/steps/step_webhook_events.go:18]: Could we implement a hybrid approach where the user has to manually trigger a webhook event by e.g. running `/zoom start` and we then problematically verify that a webhook event was revived.

File: server/http.go
Comment (hanzei) [reply server/http.go:76]: Makes sense. Thanks for the explanation :+1:

File: server/wizard/steps/step_webhook_events.go
Comment (mickmister) [reply server/wizard/steps/step_webhook_events.go:18]: @hanzei Something like:

> To test your connection with zoom, run  `/zoom start` to start a meeting, or click the Zoom icon in the channel header.

Then in the output, we include whether the webhook was successfully verified or not.

File: server/wizard/steps/step_greeting.go
Comment (mickmister) [reply server/wizard/steps/step_greeting.go:29]: Implemented

File: server/config/config.go
Comment (mickmister) [reply server/config/config.go:42]: @hanzei Does that mess with an existing config though? I would think that changing the casing in the config's schema would be a breaking change.

**Comment by mickmister** [comment]
Here are some known remaining issues

Bugs:
- [ ] Images are not showing fully until you click them to show in a modal. Maybe we need to use slack attachment images

UX-related requests:
- [ ] From Figma design: Wording changes and step consolidation. Merging two steps into one etc.
- [ ] Delegate teammate feature
- [ ] Announcement feature
- [ ] If the user runs the command in a channel other than the Zoom bot DM channel, it should respond with an ephemeral message pointing to the DM channel.
- [x] "you can set up later" step
- [x] "cancel setup" button on all relevant steps
- [x] use interactive dialog to avoid persisting webhook secret in permanent post

Dialogue-related requests:
- [ ] Tell admin they're creating a user level app
- [ ] Include "Add scopes" in instruction wording
- [ ] We should say to ignore "Verification Code" on the webhook event page

Dev requests:

- [ ] Use struct instead of map for config
- [ ] Add more telemetry data for flow start/end. Currently just capturing the current time (definitely can be made a separate ticket)

**Comment by mickmister** [comment]
@aaron @hanzei @DHaussermann please let me know your thoughts on my comment above

**Comment by mickmister** [comment]
cc @aaronrothschild on my comment above

**Comment by mickmister** [comment]
I've made it so we hide the webhook URL from posts using a dialog, and implemented the announcement step.

**Comment by mickmister** [comment]
@DHaussermann The image sizes are fixed now. Still working on other UX-related issues

File: server/config/config.go
Comment (mickmister) [reply server/config/config.go:51]: The `mattermst-plugin-api` method `client.Configuration.SavePluginConfig` accepts a `map[string]interface{}` as its argument. How do you get around this requirement?

Comment (mickmister) [reply server/config/config.go:51]: @hanzei The `mattermst-plugin-api` method `client.Configuration.SavePluginConfig` accepts a `map[string]interface{}` as its argument. How do you get around this requirement?

**Comment by mickmister** [comment]
@DHaussermann This is ready for testing. The main thing not completed at this point is the delegate feature. Working on that now.

Could you please test the path that includes setting the vanity URL in your testing?

File: go.mod
Comment (larkox) [inline go.mod:16]: How long will this be needed? Should we create a ticket to remove it?

**Review by larkox — APPROVED** [review]

**Comment by mickmister** [comment]
@hanzei @catalintomai @aaronrothschild @DHaussermann

There are two main review requests that are not implemented at this point:

- Delegating setup to another teammate
- Automatically verifying the webhook is set up correctly during the setup flow

Also another point from @hanzei:

-  Use struct instead of map for config. See https://github.com/mattermost/mattermost-plugin-zoom/pull/249#discussion_r817288800

Is this feature good to merge at this point? If the features are considered low priority, we can make the remaining items into HW tickets.

**Comment by catalintomai** [comment]
@mickmister , I'd say merge it, pending @DHaussermann 's approval.

File: server/config/config.go
Comment (hanzei) [reply server/config/config.go:51]: Json.Unmarshal and then json.Marshal.  That is the the way the other three plugins are following.

File: server/config/config.go
Comment (hanzei) [reply server/config/config.go:42]: The existing configurations use lower case anyway as the server auto converts to lower case. Only if plugins to fancy stuff like writing there own config is when one needs to be careful.

File: server/wizard/wizard.go
Comment (hanzei) [inline server/wizard/wizard.go:105]: Could the state contain confidential information?

**Review by hanzei — APPROVED** [review]
I'm fine with deferring the outstanding features to anther ticket.

**Comment by DHaussermann** [comment]
@mickmister @dipak-demansol was seeing and intermittent issue that was happening when he was trying to open the modal. This was preventing the modal from opening. This same error would occur preventing the values from saving. 


Strangely - the error only seem related to post more generally and not Zoom?
Any high level thoughts what may have been happening here? Maybe a broader issue with plugin dialogues?
image
image

**Comment by mickmister** [comment]
@DHaussermann I've seen similar issues when testing the flow, but I'm not sure what the cause is.

@hanzei @levb Did you ever experience this during development of your flows? Intermittently, clicking a button that is meant to open an interactive dialog results in a 400, and does not open the dialog.

**Comment by hanzei** [comment]
I don't remember encountering this issue. 

Is there anything meaningful in the server logs?

**Comment by mickmister** [comment]
@hanzei @DHaussermann I consistently see this error if I don't have my ngrok tunnel up, as this is my Site URL. I recall seeing it intermittently with the ngrok tunnel active.

In the case where I'm able to reproduce by disabling the ngrok tunnel, the same response is returned to the client as it is in the screenshot. The server logs have little information, though in the case of the ngrok tunnel being down:

- Server logs 404 from post action call to the plugin
- Server returns 400 to client

This may not be useful, since this is a different case and may have different behavior. I tried to trace the problem down, but wasn't able to pinpoint the issue. There may be something going on with the flow library somewhere.

**Comment by hanzei** [comment]
Do the server logs point out which URL returned a 404?

**Comment by mickmister** [comment]
@hanzei No the log is pretty vague in that regard. Only the status code is shown

**Comment by DHaussermann** [comment]
There is some testing feedback on Step 4. of the wizard for the webhook. 
image

A. Can re-word 4.4 to clarify that this is a button in the post on the Mattermost side and not the Zoom UI? Sample text `Click the "Show Webhook URL" button below to show the webhook for your Mattermost server`
B. Should clicking Continue end the step 4? The user till has to perform step 4.5 so, it feel awkward a bit that the wizard has stepped forward
C. Step 4.6 does not exist in the UI - As far as I can tell - it's just for account level not user level

@mickmister A. and B. could be considered nice to have for a future release. But C. must be addressed.

File: go.mod
Comment (hanzei) [inline go.mod:16]: This can be dropped now

**Pull request closed without merging**
